File: app/src/main/java/com/xn1ch1/qrscavengerhunt/ActivityHunt.java
package com.xn1ch1.qrscavengerhunt;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.media.MediaPlayer;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.TextView;

public class ActivityHunt extends AppCompatActivity {

    final static int SCAN_BARCODE = 88;

    SharedPreferences mDataStore;
    SharedPreferences.Editor mDataStoreEditor;

    String[][] mClueData = new String[10][2];

    TextView mClueTitle;
    TextView mClueText;

    int mClueNumber = 0;

    Boolean mIsLastClue;

    MediaPlayer mErrorSound;
    MediaPlayer mSolvedSound;
    MediaPlayer mFinishSound;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_hunt);

        loadClueData();

        mErrorSound = MediaPlayer.create(this, R.raw.hunt_error);
        mSolvedSound = MediaPlayer.create(this, R.raw.hunt_solved);
        mFinishSound = MediaPlayer.create(this, R.raw.hunt_finish);

        mIsLastClue = (mClueNumber == 9 || mClueData[mClueNumber + 1][0].equals(""));

        mClueText = findViewById(R.id.huntClueText);
        mClueTitle = findViewById(R.id.huntClueTitle);

        applyClueData();
        solvedInvisible();
        finishInvisible();
        errorInvisible();

    }

    @Override
    public void onBackPressed() {
        Intent intent = new Intent(this, ActivityHome.class);
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
        startActivity(intent);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {

        if (resultCode == Activity.RESULT_OK) {

            if (requestCode == SCAN_BARCODE) {
                if (data.getStringExtra("code").equals(mClueData[mClueNumber][1])) {

                    clueInvisible();
                    errorInvisible();

                    if (mIsLastClue) {

                        writeHuntFinished();
                        mFinishSound.start();
                        finishVisible();

                        findViewById(R.id.huntLayout).setOnClickListener(new View.OnClickListener() {
                            @Override
                            public void onClick(View v) {
                                onBackPressed();
                            }
                        });

                    } else {

                        writeClueSolved();
                        mSolvedSound.start();
                        solvedVisible();

                    }

                } else {

                    mErrorSound.start();
                    errorVisible();

                }
            }

        }
    }

    /* *******************
        User Interactions
       ******************* */
    public void scanBarcode(View v) {
        errorInvisible();
        Intent scan = new Intent(this, BarcodeScan.class);
        startActivityForResult(scan, SCAN_BARCODE);
    }

    public void goBack(View view) {
        onBackPressed();
    }

    public void goNext(View view) {
        mClueNumber++;
        mIsLastClue = (mClueNumber == 9 || mClueData[mClueNumber + 1][0].equals(""));
        applyClueData();
        solvedInvisible();
        clueVisible();
        errorInvisible();
    }

    /* *****************
        Display Changes
       ***************** */
    private void applyClueData() {
        mClueText.setText(mClueData[mClueNumber][0]);
        if (mIsLastClue) {
            mClueTitle.setText(getResources().getString(R.string.huntFinalClueTitle));
        } else {
            mClueTitle.setText(getResources().getString(R.string.allClueNumberTitle, mClueNumber + 1));
        }
    }

    private void clueVisible() {
        findViewById(R.id.huntClueTitle).setVisibility(View.VISIBLE);
        findViewById(R.id.huntClueText).setVisibility(View.VISIBLE);
        findViewById(R.id.huntClueError).setVisibility(View.VISIBLE);
        findViewById(R.id.huntClueScan).setVisibility(View.VISIBLE);
        findViewById(R.id.huntClueBack).setVisibility(View.VISIBLE);
    }

    private void clueInvisible() {
        findViewById(R.id.huntClueTitle).setVisibility(View.INVISIBLE);
        findViewById(R.id.huntClueText).setVisibility(View.INVISIBLE);
        findViewById(R.id.huntClueError).setVisibility(View.INVISIBLE);
        findViewById(R.id.huntClueScan).setVisibility(View.INVISIBLE);
        findViewById(R.id.huntClueBack).setVisibility(View.INVISIBLE);
    }

    private void errorVisible() {
        findViewById(R.id.huntClueError).setVisibility(View.VISIBLE);
    }

    private void errorInvisible() {
        findViewById(R.id.huntClueError).setVisibility(View.INVISIBLE);
    }

    private void solvedVisible() {
        findViewById(R.id.huntSolvedTitle).setVisibility(View.VISIBLE);
        findViewById(R.id.huntSolvedImage).setVisibility(View.VISIBLE);
        findViewById(R.id.huntSolvedBack).setVisibility(View.VISIBLE);
        findViewById(R.id.huntSolvedNext).setVisibility(View.VISIBLE);
    }

    private void solvedInvisible() {
        findViewById(R.id.huntSolvedTitle).setVisibility(View.INVISIBLE);
        findViewById(R.id.huntSolvedImage).setVisibility(View.INVISIBLE);
        findViewById(R.id.huntSolvedBack).setVisibility(View.INVISIBLE);
        findViewById(R.id.huntSolvedNext).setVisibility(View.INVISIBLE);
    }

    private void finishVisible() {
        findViewById(R.id.huntFinishedImage).setVisibility(View.VISIBLE);
        findViewById(R.id.huntFinishedTitle).setVisibility(View.VISIBLE);
        findViewById(R.id.huntFinishedMessage).setVisibility(View.VISIBLE);
    }

    private void finishInvisible() {
        findViewById(R.id.huntFinishedImage).setVisibility(View.INVISIBLE);
        findViewById(R.id.huntFinishedTitle).setVisibility(View.INVISIBLE);
        findViewById(R.id.huntFinishedMessage).setVisibility(View.INVISIBLE);
    }

    /* *****************
        Data Read/Write
       ***************** */
    private void loadClueData() {
        mDataStore = getSharedPreferences(getResources().getString(R.string.dataClueData), Context.MODE_PRIVATE);
        mClueNumber = mDataStore.getInt(getResources().getString(R.string.dataCurrentClue), 0);
        for (int i = 0; i < 10; i++) {
            mClueData[i][0] = mDataStore.getString("clue" + i, "");
            mClueData[i][1] = mDataStore.getString("code" + i, "");
        }
    }

    public void writeClueSolved() {
        mDataStoreEditor = mDataStore.edit();
        mDataStoreEditor.putInt(getResources().getString(R.string.dataCurrentClue), mClueNumber);
        mDataStoreEditor.apply();
    }

    private void writeHuntFinished() {
        mDataStoreEditor = mDataStore.edit();
        mDataStoreEditor.putInt(getResources().getString(R.string.dataCurrentClue), 0);
        mDataStoreEditor.putBoolean(getResources().getString(R.string.dataHuntStarted), false);
        mDataStoreEditor.apply();
    }
}

File: app/src/main/java/com/xn1ch1/qrscavengerhunt/ActivityHunt.kt
package com.xn1ch1.qrscavengerhunt

import android.Manifest
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.media.MediaPlayer
import android.net.Uri
import android.os.Bundle
import android.provider.Settings
import android.util.Log
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat

class ActivityHunt : BaseActivity() {

    private lateinit var scan: ImageView
    private lateinit var back: ImageView
    private lateinit var next: ImageView

    private lateinit var title: TextView
    private lateinit var clueText: TextView
    private lateinit var clueError: TextView

    private lateinit var errorSound: MediaPlayer
    private lateinit var solvedSound: MediaPlayer
    private lateinit var finishedSound: MediaPlayer

    private lateinit var cluePrefs: SharedPreferences

    private var clue: Clue = Clue()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_hunt)

        cluePrefs = this.getSharedPreferences(Prefs.PREFS_FILENAME, 0)

        clue.number = prefs.currentClue

        errorSound = MediaPlayer.create(this, R.raw.hunt_error)
        solvedSound = MediaPlayer.create(this, R.raw.hunt_solved)
        finishedSound = MediaPlayer.create(this, R.raw.hunt_finish)

        scan = findViewById(R.id.huntClueScan)
        back = findViewById(R.id.huntClueBack)
        next = findViewById(R.id.huntSolvedNext)

        clueText = findViewById(R.id.huntClueText)
        title = findViewById(R.id.huntClueTitle)
        clueError = findViewById(R.id.huntClueError)

        clueText.text = clue.clueText
        title.text = getString(R.string.clueNumber, prefs.currentClue + 1)

        scan.setOnClickListener {
            checkCameraPerms()
        }
        back.setOnClickListener {
            finish()
        }
        next.setOnClickListener {
            nextClue()
        }

        hideErrorLayout()
        hideFinishedLayout()
        hideSolvedLayout()
        showClueLayout()

//        prefs.huntStarted = false
//        prefs.currentClue = 0

        if (prefs.clueCount == prefs.currentClue) {
            prefs.huntStarted = false
            prefs.currentClue = 0
            finish()
        }

    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {

        if (resultCode == RESULT_OK) {
            if (requestCode == CAMERA_PERMISSION_REQUEST_CODE) {
                if (ContextCompat.checkSelfPermission(
                                this,
                                Manifest.permission.CAMERA
                        ) == PackageManager.PERMISSION_GRANTED
                ) {
                    openCameraWithScanner()
                }
            } else if (requestCode == ACT_BARCODE) {

                hideClueLayout()

                val code = data?.getStringExtra("barcode").toString()

                if (code == clue.code) {

                    if (prefs.clueCount == prefs.currentClue + 1) {

                        prefs.huntStarted = false
                        prefs.currentClue = 0

                        showFinishedLayout()
                        finishedSound.start()

                    } else {

                        prefs.currentClue = prefs.currentClue + 1

                        showSolvedLayout()
                        solvedSound.start()

                    }

                } else {

                    showErrorLayout()
                    errorSound.start()

                }

            }
        }

    }


    override fun onRequestPermissionsResult(
            requestCode: Int,
            permissions: Array<out String>,
            grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)

        if (requestCode == CAMERA_PERMISSION_REQUEST_CODE && grantResults.isNotEmpty()) {
            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                openCameraWithScanner()
            } else if (!ActivityCompat.shouldShowRequestPermissionRationale(
                            this,
                            Manifest.permission.CAMERA
                    )
            ) {
                val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS)
                val uri: Uri = Uri.fromParts("package", packageName, null)
                intent.data = uri
                startActivityForResult(intent, CAMERA_PERMISSION_REQUEST_CODE)
            }
        }

    }

    private fun openCameraWithScanner() {

        val intent = Intent(this, ActivityScanCode::class.java)
        startActivityForResult(intent, ACT_BARCODE)

    }

    private fun checkCameraPerms() {

        if (ContextCompat.checkSelfPermission(
                        this,
                        Manifest.permission.CAMERA
                ) == PackageManager.PERMISSION_GRANTED
        ) {
            openCameraWithScanner()
        } else {
            ActivityCompat.requestPermissions(
                    this,
                    arrayOf(Manifest.permission.CAMERA),
                    CAMERA_PERMISSION_REQUEST_CODE
            )
        }

    }

    private fun nextClue() {
        recreate()
    }


    private fun showClueLayout() {
        title.visibility = View.VISIBLE
        clueText.visibility = View.VISIBLE
        scan.visibility = View.VISIBLE
        back.visibility = View.VISIBLE
    }

    private fun hideClueLayout() {
        title.visibility = View.INVISIBLE
        clueText.visibility = View.INVISIBLE
        scan.visibility = View.INVISIBLE
        back.visibility = View.INVISIBLE
    }

    private fun showErrorLayout() {
        clueError.visibility = View.VISIBLE
        back.visibility = View.VISIBLE
    }

    private fun hideErrorLayout() {
        clueError.visibility = View.INVISIBLE
        back.visibility = View.INVISIBLE
    }

    private fun showSolvedLayout() {
        findViewById<View>(R.id.huntSolvedTitle).visibility = View.VISIBLE
        findViewById<View>(R.id.huntSolvedImage).visibility = View.VISIBLE
        back.visibility = View.VISIBLE
        next.visibility = View.VISIBLE
    }

    private fun hideSolvedLayout() {
        findViewById<View>(R.id.huntSolvedTitle).visibility = View.INVISIBLE
        findViewById<View>(R.id.huntSolvedImage).visibility = View.INVISIBLE
        back.visibility = View.INVISIBLE
        next.visibility = View.INVISIBLE
    }

    private fun showFinishedLayout() {
        findViewById<View>(R.id.huntFinishedImage).visibility = View.VISIBLE
        findViewById<View>(R.id.huntFinishedTitle).visibility = View.VISIBLE
        findViewById<View>(R.id.huntFinishedMessage).visibility = View.VISIBLE
        findViewById<ConstraintLayout>(R.id.huntLayout).setOnClickListener {
            finish()
        }
    }

    private fun hideFinishedLayout() {
        findViewById<View>(R.id.huntFinishedImage).visibility = View.INVISIBLE
        findViewById<View>(R.id.huntFinishedTitle).visibility = View.INVISIBLE
        findViewById<View>(R.id.huntFinishedMessage).visibility = View.INVISIBLE
    }

    inner class Clue {

        var number: Int = 0

        var clueText: String?
            get() = cluePrefs.getString("clueText$number", "")
            set(value) {
                cluePrefs.edit().putString("clueText$number", value).apply()
            }

        var code: String?
            get() = cluePrefs.getString("code$number", "")
            set(value) {
                cluePrefs.edit().putString("code$number", value).apply()
            }

    }

}